include/bbp/sonata/compartment_sets.h

```diff
@@ -0,0 +1,169 @@
+    explicit CompartmentLocation(std::unique_ptr<detail::CompartmentLocation>&& impl);
+    CompartmentLocation(const CompartmentLocation& other);
+    CompartmentLocation& operator=(const CompartmentLocation& other);
+    CompartmentLocation(CompartmentLocation&&) noexcept;
+    CompartmentLocation& operator=(CompartmentLocation&&) noexcept;
+    ~CompartmentLocation();
+
+    bool operator==(const CompartmentLocation& other) const noexcept;
+    bool operator!=(const CompartmentLocation& other) const noexcept;
+
+    uint64_t gid() const;
+    uint64_t sectionIdx() const;
+    double offset() const;
+
+    std::string toJSON() const;
+
+  private:
+    std::unique_ptr<detail::CompartmentLocation> impl_;
+};
+
+class SONATA_API CompartmentSetFilteredIterator {
+public:
+    using iterator_category = std::input_iterator_tag;
+    using value_type = CompartmentLocation;
+    using difference_type = std::ptrdiff_t;
+    using pointer = void;                // dereference returns by value
+    using reference = CompartmentLocation; // dereference returns by value
+
+    explicit CompartmentSetFilteredIterator(std::unique_ptr<detail::CompartmentSetFilteredIterator> impl);
+    CompartmentSetFilteredIterator(const CompartmentSetFilteredIterator& other);
+    CompartmentSetFilteredIterator& operator=(const CompartmentSetFilteredIterator& other);
+    CompartmentSetFilteredIterator(CompartmentSetFilteredIterator&&) noexcept;
+    CompartmentSetFilteredIterator& operator=(CompartmentSetFilteredIterator&&) noexcept;
+    ~CompartmentSetFilteredIterator();
+
+    /// Dereference operator. It makes a copy!
+    CompartmentLocation operator*() const;
+    /// Arrow operator is voluntarely disabled because we can only return copies of CompartmentLocation.
+    /// In any way we need to find a location to store a temp CompartmentLocation and memory leaks become possible.
+    CompartmentSetFilteredIterator& operator++();            // prefix ++
+    CompartmentSetFilteredIterator operator++(int);          // postfix ++
+    bool operator==(const CompartmentSetFilteredIterator& other) const;
+    bool operator!=(const CompartmentSetFilteredIterator& other) const;
+
+private:
+    std::unique_ptr<detail::CompartmentSetFilteredIterator> impl_;
+};
+
+/**
+ * CompartmentSet public API.
+ *
+ * This class represents a set of compartment locations associated with a neuron population.
+ * Each compartment is uniquely defined by a (gid, section_idx, offset) triplet.
+ * This API supports filtering based on a gid selection.
+ */
+class SONATA_API CompartmentSet
+{
+  public:
+
+    CompartmentSet() = delete;
+
+    explicit CompartmentSet(const std::string& json_content);
+    explicit CompartmentSet(std::shared_ptr<detail::CompartmentSet>&& impl);
+
+    std::pair<CompartmentSetFilteredIterator, CompartmentSetFilteredIterator>
+    filtered_crange(bbp::sonata::Selection selection = bbp::sonata::Selection({})) const;
```

Comment:
this is all in the `bbp:sonata` namespace, so these qualifications are all unnecessary.

Reply:
done